design/KR/kenward.Rmd

```diff
@@ -0,0 +1,402 @@
+---
+title: how to implement Kenward Roger degree of freedom?
+author: "Liming Li"
+output: html_document
+editor_options:
+  chunk_output_type: console
+---
+
+# Objective
+
+The objective is to design how we can implement Kenward-Roger degree of freedom.
+
+# Current design of mmrm package
+
+In current mmrm package, we provided Satterthwaite degree of freedom, through `df_1d` and `df_md`.
+
+# Design of Kenward-Roger degree of freedom
+
+## update to current functions
+
+* `df_1d` and `df_md` renamed to `df_1d_sat` and `df_md_sat`, do not export them
+* add a new function `df_1d` and `df_md` to include degree of freedom selection
+* implement `df_1d_kr` and `df_md_kr` (take extra arguments to conduct improved Kenward-Roger degree of freedom)
+
+## Further updates to the naming
+* consider renaming the `df_1d` and `df_md` to something like `beta_test_1d` and `beta_test_md`?(the names need to be discussed)
+* add deprecation notes on `df_1d` and `df_md`
+
+## pseudo code
+
+```{r, eval=FALSE}
+# df_1d
+formula <- FEV1 ~ ARMCD + ad(AVISIT | USUBJID)
+result <- mmrm(formula, fev_data, reml = FALSE)
+df_1d(result, contrast = c(0, 1), df = "SAT") # test the armcd using SAT
+# df_1d_sat is called
+df_1d(result, contrast = c(0, 1), df = "KR") # KR defaults to KR2 (improved version)
```

Comment:
and then here if we ask for `KR` but it has not been calculated in `mmrm()` we fail